File: verify_correctness/plot.py
import matplotlib.pyplot as plt

def plot_from_txt(filename, dimen, population, max_iter):
    generations = []
    values = []

    # Read data from txt file
    try:
        with open(filename, 'r') as file:
            for line in file:
                parts = line.split()
                if len(parts) == 2:
                    generations.append(int(parts[0]))
                    values.append(float(parts[1]))
    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
        return
    except ValueError:
        print("Error: Unable to parse the data from the file. Ensure it is formatted correctly.")
        return

    # Plot the data
    plt.figure(figsize=(12, 8))
    plt.plot(generations, values, label='Best value so far', color='g', linewidth=2)
    plt.xlabel('GENERATIONS', fontsize=15)
    plt.ylabel('VALUE', fontsize=15)
    plt.title(f'Best Value Over Generations (Dimen: {dimen}, Population: {population}, Max Iter: {max_iter})', fontsize=18)
    plt.grid(True)
    plt.legend()

    # Save and show the plot
    output_file = input("Enter output image filename (e.g., output.png): ")
    plt.savefig(output_file)
    print(f"Plot saved to {output_file}")
    plt.show()

if __name__ == "__main__":
    filename = input("Enter the filename to plot data from: ")
    dimen = int(input("Enter dimension: "))
    population = int(input("Enter population: "))
    max_iter = int(input("Enter max iterations: "))
    plot_from_txt(filename, dimen, population, max_iter)

File: verify_correctness/plot.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <matplotlibcpp.h> // Make sure you have matplotlib-cpp installed and configured correctly.

namespace plt = matplotlibcpp;

int main() {
    std::string filename;
    int dimen, population, max_iter;

    // Get the parameters from the user
    std::cout << "Enter dimension, population, and max iterations: ";
    std::cin >> dimen >> population >> max_iter;

    // Get the filename from the user
    std::cout << "Enter the filename to plot data from: ";
    std::cin >> filename;

    // Read the data from the file
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Unable to open file: " << filename << std::endl;
        return 1;
    }

    std::vector<int> generations;
    std::vector<double> values;
    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        int generation;
        double value;
        if (!(iss >> generation >> value)) {
            break;
        }
        generations.push_back(generation);
        values.push_back(value);
    }
    file.close();

    // Plot the data
    plt::figure_size(1200, 780);
    plt::xlabel("Generations");
    plt::ylabel("Value");
    plt::title("Best Value Over Generations (Dimen: " + std::to_string(dimen) + ", Population: " + std::to_string(population) + ", Max Iter: " + std::to_string(max_iter) + ")");
    plt::plot(generations, values, "g-");
    plt::grid(true);

    // Save or show the plot
    std::string output_file;
    std::cout << "Enter output image filename (e.g., output.png): ";
    std::cin >> output_file;
    plt::save(output_file);
    std::cout << "Plot saved to " << output_file << std::endl;

    return 0;
}
